File: game/entities/win_dialog.go
package entities

import (
	"fmt"
	"github.com/kraxarn/ubongo/res"
	"image"
	"math/rand"
	"time"
)

type WinDialog struct {
	*Dialog
}

func NewWinDialog(total time.Duration, screenSize image.Point) (*WinDialog, error) {
	dialog, err := newDialog(screenSize)
	if err != nil {
		return nil, err
	}

	dialog.setTitle(randomMessage(total))
	dialog.setInfo(fmt.Sprintf("You completed in %.3f seconds", total.Seconds()))

	dialog.addLeftButton(res.Return, "Back to menu")
	dialog.addRightButton(res.Forward, "Next level")

	return &WinDialog{
		Dialog: dialog,
	}, nil
}

func (d *WinDialog) SetOnBack(pressed func()) {
	d.onLeftPressed = pressed
}

func (d *WinDialog) SetOnNext(pressed func()) {
	d.onRightPressed = pressed
}

func randomMessage(total time.Duration) string {
	var messages []string

	if total < time.Second*10 {
		messages = []string{
			"Amazing!",
			"Incredible!",
			"That was quick!",
		}
	} else if total > time.Minute {
		messages = []string{
			"neat",
			"Puzzle Defeated",
			"Just in time!",
			"Phew!",
		}
	} else {
		messages = []string{
			"You won!",
			"You did it!",
			"Cool!",
			"Congratulations!",
		}
	}

	// Seed should be reset when loading the next level,
	// so it should be safe to change it here
	rand.Seed(time.Now().Unix())
	return messages[rand.Intn(len(messages))]
}

File: src/commonMain/kotlin/containers/WinDialog.kt
package containers

import Resources
import com.soywiz.klock.TimeSpan
import com.soywiz.klock.minutes
import com.soywiz.klock.seconds
import com.soywiz.korge.input.onClick
import com.soywiz.korge.view.Container
import com.soywiz.korge.view.ViewDslMarker
import com.soywiz.korge.view.addTo
import com.soywiz.korim.vector.format.SVG
import com.soywiz.korim.vector.render
import com.soywiz.korio.util.toStringDecimal
import enums.ResFont
import enums.ResImage

fun Container.winDialog(
	res: Resources,
	total: TimeSpan,
	width: Double, height: Double,
	callback: @ViewDslMarker WinDialog.() -> Unit = {},
) = WinDialog(res, total, width, height).addTo(this, callback)

class WinDialog(
	res: Resources,
	total: TimeSpan,
	width: Double, height: Double,
) : Dialog(width, height, res[ResFont.BOLD], res[ResFont.REGULAR])
{
	val onBack = onLeftAction
	val onNext = onRightAction

	init
	{
		title(randomMessage(total))
		body("You completed in ${total.seconds.toStringDecimal(3)} seconds")

		val backIcon = SVG(res[ResImage.UI_ARROW_LEFT]).render()
		val nextIcon = SVG(res[ResImage.UI_PLAY]).render()

		leftAction("Back to menu", backIcon) {
			onClick { onBack(it) }
		}

		rightAction("Next level", nextIcon) {
			onClick { onNext(it) }
		}
	}

	private fun randomMessage(total: TimeSpan) = (when
	{
		total < 10.seconds -> arrayOf("Amazing!", "Incredible!", "That was quick!")
		total > 1.minutes -> arrayOf("neat", "Puzzle Defeated", "Just in time!", "Phew!")
		else -> arrayOf("You won!", "You did it!", "Cool!", "Congratulations!")
	}).random()
}